launch/shell.go

```diff
@@ -46,44 +45,10 @@ func (l *Launcher) launchWithShell(self string, process Process) error {
 
 func (l *Launcher) profiles(process Process) ([]string, error) {
 	var profiles []string
-
-	appendIfFile := func(path string, fi os.FileInfo) {
-		if !fi.IsDir() {
-			profiles = append(profiles, path)
-		}
-	}
-
-	appendFilesInDir := func(path string) error {
-		fis, err := ioutil.ReadDir(path)
-		if os.IsNotExist(err) {
-			return nil
-		}
-		if err != nil {
-			return errors.Wrapf(err, "failed to list files in dir '%s'", path)
-		}
-
-		for _, fi := range fis {
-			appendIfFile(filepath.Join(path, fi.Name()), fi)
-		}
-		return nil
-	}
-
 	if err := l.eachBuildpack(func(path string) error {
-		absPath, err := filepath.Abs(path)
-		if err != nil {
-			return err
-		}
-		return eachDir(absPath, func(path string) error {
-			if err := appendFilesInDir(filepath.Join(path, "profile.d")); err != nil {
-				return err
-			}
-			if process.Type != "" {
-				return appendFilesInDir(filepath.Join(path, "profile.d", process.Type))
-			}
-			return nil
-		})
+		return eachLayer(path, l.populateLayerProfiles(process, &profiles))
```

Comment:
Having trouble leaving a code suggestion, but path -> bpDir might be clearer here as well.